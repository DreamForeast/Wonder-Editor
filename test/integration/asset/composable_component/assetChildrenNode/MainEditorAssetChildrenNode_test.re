open Wonder_jest;

open Expect;

open Expect.Operators;

open CurrentNodeDataType;

open Sinon;

open AssetNodeType;

let _ =
  describe("MainEditorAssetChildrenNode", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initStateAndGl(~sandbox, ());
      MainEditorSceneTool.createDefaultScene(
        sandbox,
        MainEditorAssetTool.initAssetTree,
      );
      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });
    afterEach(() => {
      restoreSandbox(refJsObjToSandbox(sandbox^));
      StateAssetService.getState()
      |> CurrentNodeDataAssetService.clearCurrentNodeData
      |> CurrentNodeParentIdAssetService.clearCurrentNodeParentId
      |> StateAssetService.setState
      |> ignore;
    });

    describe("test set current node", () => {
      test("click texture file to be current node", () => {
        let assetTreeDomRecord =
          MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

        assetTreeDomRecord
        |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstTextureDomIndex
        |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

        let {currentNodeId, nodeType} =
          StateAssetService.getState()
          |> CurrentNodeDataAssetService.unsafeGetCurrentNodeData;

        (currentNodeId, nodeType)
        |>
        expect == (
                    assetTreeDomRecord
                    |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstTextureNodeId,
                    AssetNodeType.Texture,
                  );
      });

      test("click json file to be current node", () => {
        let assetTreeDomRecord =
          MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

        assetTreeDomRecord
        |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstJsonDomIndex
        |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

        let {currentNodeId, nodeType} =
          StateAssetService.getState()
          |> CurrentNodeDataAssetService.unsafeGetCurrentNodeData;

        (currentNodeId, nodeType)
        |>
        expect == (
                    assetTreeDomRecord
                    |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstJsonNodeId,
                    AssetNodeType.Json,
                  );
      });

      describe("test click folder", () => {
        describe("test single click", () => {
          testPromise("test set folder to be current node", () => {
            let assetTreeDomRecord =
              MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
            let fakeDom =
              EventListenerTool.buildFakeDom()
              |> EventListenerTool.stubGetElementByIdReturnFakeDom;

            BuildComponentTool.buildAssetChildrenNode(10);

            EventListenerTool.triggerEvent(fakeDom, "mousedown", {});
            Js.Promise.make((~resolve, ~reject) =>
              Timeout.setTimeout(
                () => {
                  EventListenerTool.triggerEvent(fakeDom, "mousedown", {});
                  resolve(.
                    {
                      let {currentNodeId, nodeType} =
                        StateAssetService.getState()
                        |> CurrentNodeDataAssetService.unsafeGetCurrentNodeData;

                      (currentNodeId, nodeType)
                      |>
                      expect == (
                                  assetTreeDomRecord
                                  |> MainEditorAssetNodeTool.OperateTwoLayer.getSecondFolderNodeId,
                                  AssetNodeType.Folder,
                                );
                    },
                  );
                },
                20,
              )
            );
          });
          testPromise("test snapshot", () => {
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;
            let fakeDom =
              EventListenerTool.buildFakeDom()
              |> EventListenerTool.stubGetElementByIdReturnFakeDom;

            BuildComponentTool.buildAssetChildrenNode(10);

            EventListenerTool.triggerEvent(fakeDom, "mousedown", {});
            Js.Promise.make((~resolve, ~reject) =>
              Timeout.setTimeout(
                () => {
                  EventListenerTool.triggerEvent(fakeDom, "mousedown", {});
                  resolve(.
                    BuildComponentTool.buildAssetComponent()
                    |> ReactTestTool.createSnapshotAndMatch,
                  );
                },
                20,
              )
            );
          });
        });

        testPromise(
          "double click folder, set folder to be currentAssetNodeParent and currentNode(are the same)",
          () => {
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;
            let fakeDom =
              EventListenerTool.buildFakeDom()
              |> EventListenerTool.stubGetElementByIdReturnFakeDom;

            BuildComponentTool.buildAssetChildrenNode(10);

            EventListenerTool.triggerEvent(fakeDom, "mousedown", {});
            Js.Promise.make((~resolve, ~reject) =>
              Timeout.setTimeout(
                () => {
                  EventListenerTool.triggerEvent(fakeDom, "mousedown", {});
                  Timeout.setTimeout(
                    () => {
                      EventListenerTool.triggerEvent(
                        fakeDom,
                        "mousedown",
                        {},
                      );
                      resolve(.
                        {
                          let currentNodeId =
                            MainEditorAssetNodeTool.getCurrentNodeId();

                          let currentNodeParentId =
                            CurrentNodeParentIdAssetService.unsafeGetCurrentNodeParentId
                            |> StateLogicService.getAssetState;

                          currentNodeId |> expect == currentNodeParentId;
                        },
                      );
                    },
                    20,
                  );
                },
                5,
              )
            );
          },
        );
      });
    });
  });