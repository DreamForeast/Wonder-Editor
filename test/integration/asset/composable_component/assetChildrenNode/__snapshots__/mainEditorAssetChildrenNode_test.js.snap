// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorAssetChildrenNode test set current node test click folder test single click test snapshot 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="asset-header-item"
        onClick={[Function]}
      >
        <div
          className="item-canBeClick"
        >
          <img
            src="./public/img/add.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
        onClick={[Function]}
      >
        <div
          className="item-canBeClick"
        >
          <img
            src="./public/img/remove.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
      >
        <div
          className="item-canBeClick"
        >
          <img
            src="./public/img/load.png"
          />
          <input
            className="asset-fileLoad"
            multiple={false}
            onChange={[Function]}
            type="file"
          />
        </div>
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
            onMouseDown={[Function]}
          >
            <img
              src="./public/img/down.png"
            />
          </div>
          <img
            className="treeNode-icon"
            src="./public/img/package.png"
          />
          <div
            className="draggable-container select-not-active"
            draggable={true}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
          >
            Assets
          </div>
        </li>
        <ul
          className="wonder-tree-node"
        >
          <li>
            <div
              className="item-triangle"
            />
            <img
              className="treeNode-icon"
              src="./public/img/package.png"
            />
            <div
              className="draggable-container"
              draggable={true}
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              onMouseDown={[Function]}
              style={Object {}}
            >
              NoName Folder
            </div>
          </li>
        </ul>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  >
    <article
      className="wonder-asset-folderBox"
      id="folder-4"
      style={
        Object {
          "border": "0px",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <div
        className="box-image"
      >
        <img
          src="./public/img/assetPackage.png"
        />
      </div>
      <div
        className="item-text item-active"
      >
        <span>
          NoName Folder
        </span>
      </div>
    </article>
  </article>
</article>
`;

exports[`MainEditorAssetChildrenNode test show order 
        order should be:
        1)for different type_:folder,wdb,material,texture;
        2)for the same type_:sort by firstname alphabetically
         1`] = `
<article
  className="wonder-asset-assetChildren"
>
  <article
    className="wonder-asset-folderBox"
    id="folder-5"
    style={
      Object {
        "border": "0px",
      }
    }
  >
    <div
      className="item-ground"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    />
    <div
      className="box-image"
    >
      <img
        src="./public/img/assetPackage.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        AFolder
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-folderBox"
    id="folder-3"
    style={
      Object {
        "border": "0px",
      }
    }
  >
    <div
      className="item-ground"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    />
    <div
      className="box-image"
    >
      <img
        src="./public/img/assetPackage.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        FFolder
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-folderBox"
    id="folder-8"
    style={
      Object {
        "border": "0px",
      }
    }
  >
    <div
      className="item-ground"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    />
    <div
      className="box-image"
    >
      <img
        src="./public/img/assetPackage.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        Materials
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-folderBox"
    id="folder-10"
    style={
      Object {
        "border": "0px",
      }
    }
  >
    <div
      className="item-ground"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    />
    <div
      className="box-image"
    >
      <img
        src="./public/img/assetPackage.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        Textures
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/wdb.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        A_WDB
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/wdb.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        C_WDB
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/mat.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        AMaterial
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/mat.png"
      />
    </div>
    <div
      className="item-text item-active"
    >
      <span>
        CMaterial
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="newImgBase64"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        AImage
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="newImgBase64"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        BImage
      </span>
    </div>
  </article>
</article>
`;
